File: unit-test-practice-2/src/main/java/dojo/ComparableVersion.java
package dojo;

import java.util.HashSet;

/**
 * Ezt az osztályt arra szeretnénk majd használni, hogy konfigurációs fájlokban megadott
 * verziószámokkal tudjunk műveleteket megadni. Egyrészről össze szeretnénk hasonlítani
 * két verziószámot - erre való a {@link #compareTo(ComparableVersion)} metódus.
 * Másrészről az egyes verziószámokat engedélyezni vagy tiltani szeretnénk - erre való
 * a {@link #setEnabled(boolean)} ill. az {@link #isEnabled()} metódus.
 */
public class ComparableVersion implements Comparable<ComparableVersion> {

	private static HashSet<ComparableVersion> enabledVersions = new HashSet<>();
	private final int[] parts;
	
	public ComparableVersion(String version) {
		this.parts = createParts(version);
	}

	@Override
	public int compareTo(ComparableVersion otherVersion) {
		int limit = Math.min(parts.length, otherVersion.parts.length);
		for (int i=0; i < limit; i++) {
			int currentResult = parts[i] - otherVersion.parts[i];
			if (currentResult != 0) {
				return currentResult;
			}
		}
		return otherVersion.parts.length - parts.length;
	}
	
	@Override
	public boolean equals(Object otherObject) {
		return (otherObject instanceof ComparableVersion) && compareTo((ComparableVersion)otherObject) == 0;
	}
	
	/**
	 * A verziószám engedélyezett?
	 */
	public boolean isEnabled() {
		synchronized (enabledVersions) {
			return enabledVersions.contains(this);	
		}
	}

	/**
	 * Engedélyezi vagy tiltja a verziószámot.
	 * Ha a paraméter <code>true</code>: engedélyezi.
	 * Ha a paraméter <code>false</code>: tiltja.
	 */
	public void setEnabled(boolean enabled) {
		synchronized (enabledVersions) {
			if (enabled) {
				enabledVersions.add(this);
			} else {
				enabledVersions.remove(this);
			}
		}
	}
	
	private final int[] createParts(String version) {
		String[] strArray = version.split("\\.");
		int[] result = new int[strArray.length];
		for (int i=0; i < strArray.length; i++) {
			result[i] = Integer.parseInt(strArray[i]);
		}
		return result;
	}
}

File: unit-test-practice-2/src/main/kotlin/dojo/ComparableVersion.kt
package dojo

import java.util.Collections
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.min

/**
 * Ezt az osztályt arra szeretnénk majd használni, hogy konfigurációs fájlokban megadott
 * verziószámokkal tudjunk műveleteket megadni. Egyrészről össze szeretnénk hasonlítani
 * két verziószámot - erre való a [compareTo] metódus.
 * Másrészről az egyes verziószámokat engedélyezni vagy tiltani szeretnénk - erre való
 * a [setEnabled] ill. az [isEnabled] metódus.
 */
class ComparableVersion(version: String) : Comparable<ComparableVersion> {

    companion object {
        private val enabledVersions = Collections.newSetFromMap(ConcurrentHashMap<ComparableVersion, Boolean>())
    }

    private val parts: IntArray = createParts(version)

    override fun compareTo(other: ComparableVersion): Int {
        val limit = min(parts.size, other.parts.size)
        for (i in 0 until limit) {
            val currentResult = parts[i] - other.parts[i]
            if (currentResult != 0) {
                return currentResult
            }
        }
        return other.parts.size - parts.size
    }

    override fun equals(other: Any?): Boolean {
        return (other is ComparableVersion) && compareTo(other) == 0
    }

    override fun hashCode(): Int {
        return parts.contentHashCode()
    }

    /**
     * A verziószám engedélyezett?
     */
    fun isEnabled(): Boolean {
        return enabledVersions.contains(this)
    }

    /**
     * Engedélyezi vagy tiltja a verziószámot.
     * Ha a paraméter `true`: engedélyezi.
     * Ha a paraméter `false`: tiltja.
     */
    fun setEnabled(enabled: Boolean) {
        if (enabled) {
            enabledVersions.add(this)
        } else {
            enabledVersions.remove(this)
        }
    }

    private fun createParts(version: String): IntArray {
        return version.split(".").map { it.toInt() }.toIntArray()
    }
}